Snap fix vault paths

tonyespy opened the pull request:
The first change in this PR updates the initial scripts to check for a pre-existing value for **_VAULT_DIR** and if none found, set it to the default value "/vault".

The second change makes a similar change in **pki-setup.sh**, and extracts some of the logic from DockerFile.vault such that it can be shared with the snap package.

Although I've tested that the shell script work as expected, I don't have the capacity to test the DockerFile change without some assistance.

Commit:
[snap] Fix hard-coded /vault paths in scripts

The path /vault was hard-coded in a number of shell scripts,
which made it difficult to specify alternate paths required
when running EdgeX in a snap. This change allows the snap
to specify _VAULT_DIR on the command line when running the
scripts.

Signed-off-by: Tony Espy <espy@canonical.com>

anonymouse64 commented on pki-setup.sh:
```diff
@@ -187,6 +187,15 @@ if [[ ! -f $1 ]]; then
+# Variables and parameters
+echo ">> Setup _VAULT_DIR and fix permissions"
+_VAULT_DIR=${_VAULT_DIR:-/vault}
+chown -R vault:vault ${_VAULT_DIR}
```
Let's put this inside a `if` check for some `SNAP*` variable, as on Ubuntu Core we won't have a `vault` user or group. Something like
```bash
if [ -z "$SNAP" ]; then
    chown -R vault:vault ${_VAULT_DIR}
    chown -R vault:vault ${_VAULT_DIR}/pki
fi
```

anonymouse64 replied:
Also note that I meant this check to be for detecting any kind of snap install vs docker, and not to distinguish between classic and core snap environments. But I'm fine if you want to check for classic vs ubuntu core systems too.

anonymouse64 reviewed (changes requested):
Looks good to me (I also haven't tested the docker changes, so I can't speak to whether the docker packaging still works)

alain2sf reviewed (commented):
I know these scripts are a temporary implementation, and will be replaced with Golang implementation, but the pki-setup.sh function was designed (originally...) to remain agnostic of the TLS/PKI target system, to remain reusable. Meaning, it would work agnostically for a Vault server, a Kong server, an Apache Web server, etc. Therefore adding Vault specific variables (extracted from Dockerfile) will negate that very purpose. And of course the Dockerfile for Vault should not depend on the TLS setup in order to get the right ownership/permissions; as we could setup Vault without TLS. So, to summarize I'm not sure these modifications are the right way to proceed... thus my confusion here.

tonyespy commented:
@anonymouse64 Based on our earlier discussion about snaps running services as root, are any of the changes I made to pki-setup.sh really needed in the snap?  If they are, then let's just bite the bullet and make the changes locally in the snap.

@alain2sf We'll just need to make sure that when the Go replacements are designed, that little if any of the setup required should be hard-coded in the Dockerfiles.  Does that sound reasonable?

anonymouse64 commented:
If it's the case that we don't want to put vault specific code inside the pki-setup.sh script can we put the vault specific code inside a `vault-pki-setup.sh` script that the Dockerfile executes? That file would basically just have the code:
```bash
# Variables and parameters
echo ">> Setup _VAULT_DIR and fix permissions"
_VAULT_DIR=${_VAULT_DIR:-/vault}

# Don't chown in snap, as snaps don't support daemons using
# setuid/gid to drop from root to a specified user/group.
if [ -z "$SNAP" ]; then
        chown -R vault:vault ${_VAULT_DIR}
        chown -R vault:vault ${_VAULT_DIR}/pki
fi

chmod 644 ${_VAULT_DIR}/config/local.json
chmod 750 ${_VAULT_DIR}/pki
chmod 640 ${_VAULT_DIR}/pki/*/*
```

Then the snap + dockerfile will still share the same setup script and the pki-setup script is independent of vault.

alain2sf commented:
@anonymouse64 That would be indeed a good alternative to keep pki-setup.sh agnostic and reusable for other servers than Vault. We need to be careful there are 3 shell commands to proceed after the pki-setup.sh execution:

`chown -R vault:vault ${_VAULT_DIR}/pki` 
`chmod 750 ${_VAULT_DIR}/pki`
`chmod 640 ${_VAULT_DIR}/pki/*/*`

====> These chown/chmiod have to be applied after pki-setup.sh execution as the pki folder and sub-folders are generated by pki-setup.sh.

I can also have the 2 above chmod incorporated into pki-setup.sh as this pki folder is created and populated by the script (relative path). The chmod are already there but with wrong octal form, easy to change.

tonyespy commented:
OK, so sounds like what we want then is the new script as proposed by @anonymouse64 called vault-pki-setup.sh that contains the following:

```
# Variables and parameters
echo ">> Setup _VAULT_DIR and fix permissions"
_VAULT_DIR=${_VAULT_DIR:-/vault}

# Don't chown in snap, as snaps don't support daemons using
# setuid/gid to drop from root to a specified user/group.
if [ -z "$SNAP" ]; then
        chown -R vault:vault ${_VAULT_DIR}
fi

chmod 644 ${_VAULT_DIR}/config/local.json
```
And then in the pki-setup.sh script we need to add the following code after the /pki dir is created:

```
if [ "$HOST" == "edgex-vault" ]; then
        _VAULT_DIR=${_VAULT_DIR:-/vault}

        if [ -z "$SNAP" ]; then
                chown -R vault:vault ${_VAULT_DIR}/pki
        fi
        chmod 750 ${_VAULT_DIR}/pki
        chmod 640 ${_VAULT_DIR}/pki/*/*
fi



```

anonymouse64 commented:
@tonyespy I'm not sure I follow... 
1) Wouldn't the `chmod`'s not get executed in your second snippet if we're inside a snap (I don't think we can rely on `$HOST == "edgex-vault"` inside any snap)? I was under the impression that on the snap the only thing we want to drop is the `chown` calls as we won't have that vault user.
2) I think it would be simpler if we made the new vault-pki-setup.sh script basically do everything we need it to:
```
# Variables and parameters
echo ">> Setup _VAULT_DIR and fix permissions"
_VAULT_DIR=${_VAULT_DIR:-/vault}

# Don't chown in snap, as snaps don't support daemons using
# setuid/gid to drop from root to a specified user/group.
if [ -z "$SNAP" ]; then
        chown -R vault:vault ${_VAULT_DIR}
fi

chmod 644 ${_VAULT_DIR}/config/local.json

# Call the pki-setup script
${SCRIPT_DIR}/pki-setup.sh pki-config.env

# Cleanup
# Don't chown things on snap - the vault user isn't guaranteed to exist
if [ -z "$SNAP" ]; then
    chown -R vault:vault ${_VAULT_DIR}/pki
fi
chmod 750 ${_VAULT_DIR}/pki
chmod 640 ${_VAULT_DIR}/pki/*/*
```

Then both the snap and the dockerfile call this one setup script, and the pki-setup script can stay as generalized as it needs to.

Commit:
[snap] Extract vault setup logic from Dockerfile

This change extracts some of the setup logic in
Dockerfile.vault and places it in pki-setup.sh.
This allows the snap to use the same logic that
previously was hard-coded in the Dockerfile.

Signed-off-by: Tony Espy <espy@canonical.com>

anonymouse64 commented on vault-kong.sh:
```diff
@@ -69,7 +69,7 @@ houseKeeping # temp files and payloads
 if [[ (! -f ${_KONG_PEM}) || (! -f ${_KONG_SK}) ]]; then
     echo ">> (3) Create PKI materials for Kong TLS server certificate"
-    /vault/pki-setup.sh /vault/pki-setup-config-kong.env
+    ${_VAULT_DIR}/pki-setup.sh ${_VAULT_DIR}/pki-setup-config-kong.env
```
Now that I think about it, considering someone might eventually want to change the `pki-setup-config-kong.env` file, can we have that as a separate environment variable? Then that file can live inside `$SNAP_DATA` for the snap and be modifiable by users.

anonymouse64 commented on vault-setup.sh:
```diff
@@ -0,0 +1,38 @@
+echo ">> Setup _VAULT_DIR and fix permissions"
+_VAULT_DIR=${_VAULT_DIR:-/vault}
+
+./pki-setup.sh pki-setup-config-vault.env
```
Same comment as above, can we put `pki-setup-config-vault.env` behind an environment variable?

tonyespy replied:
If we already control the directory the file is read from, why should the filename make a difference? The snap could copy the .env files to $SNAP_DATA (or some subdir under it) where they could be modified if need be? Then the snaps just needs to set _VAULT_DIR to $SNAP_DATA.

anonymouse64 reviewed (commented):
The layout of the scripts looks good, but I noticed that the locations of the configuration files are assumed to be in the same place which means either we would have to copy the scripts that run inside `$SNAP_DATA`, or as I suggest we make the location of the configuration files a different environment variable that inside the snap we can point to `$SNAP_DATA`

tonyespy commented:
Please disregard this branch for now, as it appears there was some major re-factoring after the California branch was cut... Also with the plan to replace most of these scripts with Go services, it's unclear whether it makes sense to make these changes in master.  To be discussed at the next security WG.  I will propose a second PR with the same changes against the california branch, where we definitely want these changes to land.

anonymouse64 commented:
I think this PR is subsumed by #18 and can be closed.

anonymouse64 commented:
@tonyespy please close this

tonyespy commented:
Closing as this is obsolete...
